src/shapes/group.class.js

```diff
@@ -15,280 +21,318 @@
     },
 
     /**
-     * Removes an object from a group; Then recalculates group's dimension, position.
-     * @param {Object} object
-     * @return {fabric.Group} thisArg
-     * @chainable
+     * keeps track of the selected objects
+     * @private
      */
-    removeWithUpdate: function(object) {
-      this._restoreObjectsState();
-      fabric.util.resetObjectTransform(this);
+    __objectSelectionMonitor: function (selected, opt) {
+      var object = opt.target;
+      if (selected) {
+        this._activeObjects.push(object);
+        this._set('dirty', true);
+      }
+      else if (this._activeObjects.length > 0) {
+        var index = this._activeObjects.indexOf(object);
+        if (index > -1) {
+          this._activeObjects.splice(index, 1);
+          this._set('dirty', true);
+        }
+      }
+    },
 
-      this.remove(object);
-      this._calcBounds();
-      this._updateObjectsCoords();
-      this.setCoords();
-      this.dirty = true;
-      return this;
+    /**
+     * @private
+     * @param {boolean} watch
+     * @param {fabric.Object} object
+     */
+    _watchObject: function (watch, object) {
+      var directive = watch ? 'on' : 'off';
+      //  make sure we listen only once
+      watch && this._watchObject(false, object);
+      // TODO check what changed ischeck
+      object[directive]('changed', this.__objectMonitor);
+      object[directive]('modified', this.__objectMonitor);
+      object[directive]('selected', this.__objectSelectionTracker);
+      object[directive]('deselected', this.__objectSelectionDisposer);
     },
 
     /**
      * @private
-     */
-    _onObjectAdded: function(object) {
-      this.dirty = true;
-      object.group = this;
+     * @param {fabric.Object} object
+     * @param {boolean} [removeParentTransform] true if object is in canvas coordinate plane
+     */
+    enterGroup: function (object, removeParentTransform) {
+      if (object.group) {
+        if (object.group === this) {
+          /* _DEV_MODE_START_ */
+          console.warn('fabric.Group: duplicate objects are not supported inside group, this call has no effect');
+          /* _DEV_MODE_END_ */
+          return;
+        }
+        object.group.remove(object);
+      }
+      // can be this converted to utils?
+      if (removeParentTransform) {
+        applyTransformToObject(
+          object,
+          multiplyTransformMatrices(
+            invertTransform(this.calcTransformMatrix()),
+            object.calcTransformMatrix()
+          )
+        );
+      }
+      object.setCoords();
+      object._set('group', this);
       object._set('canvas', this.canvas);
+      this._watchObject(true, object);
+      var activeObject = this.canvas && this.canvas.getActiveObject && this.canvas.getActiveObject();
+      // if we are adding the activeObject in a group
+      if (activeObject && (activeObject === object || (object.contains && object.contains(activeObject)))) {
```

Comment:
why did you change this? bad perf

Reply:
not only is it bad perf but it is wrong as well because it checks only 1 level of nesting

Reply:
because i did not import the other function yet. the whole descendant of is not in yet

Reply:
`deep` true as we discussed or simply import isDescendant of 
I would add a TODO comment so it's visible this is not complete.

Reply:
Added a TODO comment 7d793b37